sparse_attention_hub/sparse_attention/research_attention/maskers/fixed/implementations/quest_top_k.py

```diff
@@ -0,0 +1,272 @@
+        **kwargs: Dict[str, Any],
+    ) -> Mask:
+        """Add Quest page-Top-K sparse mask."""
+        if previous_mask.is_full_mask():
+            return previous_mask
+
+        dims: AttentionTensorDimensions = self._extract_tensor_dimensions(keys, queries)
+
+        effective_heavy_size: int = self._calculate_effective_heavy_size(
+            dims.seq_len_keys
+        )
+
+        if self._should_use_full_attention(dims, effective_heavy_size):
+            return self._create_full_mask(
+                dims, previous_mask.dtype, previous_mask.device
+            )
+
+        quest_mask = self._create_quest_page_topk_mask(
+            dims,
+            effective_heavy_size,
+            keys,
+            queries,
+            attention_mask,
+            previous_mask,
+        )
+        return previous_mask.merge_mask(quest_mask, inplace=False)
+
+    def _create_quest_page_topk_mask(
+        self,
+        dims: AttentionTensorDimensions,
+        heavy_tokens: int,
+        keys: torch.Tensor,
+        queries: torch.Tensor,
+        attention_mask: torch.Tensor,
+        previous_mask: Mask,
+    ) -> Mask:
+        """
+        1) Repeat KV to match query heads (GQA/MQA -> MHA).
+        2) Page the keys, compute per-page min/max: [B,H,P,D].
+        3) Score pages with Quest bound using queries: [B,H,Q,P].
+        4) Select Top-K pages where K_pages = ceil(heavy_tokens / page_size).
+        5) Activate all tokens in selected pages, AND with attention_mask.
+        """
+        ngroups = _get_num_key_value_groups(queries, keys)
+        keys_rep = repeat_kv(keys, ngroups)
+
+        B, H, K, D = keys_rep.shape
+        _, _, Q, _ = queries.shape
+
+        page_size = self.page_size
+        num_pages = (K + page_size - 1) // page_size
+
+        # Step 2.1: per-page min/max (vectorized)
+        page_min, page_max = self._compute_page_min_max(
+            keys_rep, page_size, num_pages
+        )
+
+        # Step 2.2: Quest scores
+        page_scores = self._quest_page_scores(queries, page_min, page_max)
+
+        # Respect external attention mask by forbidding pages with no valid tokens
+        if attention_mask is not None:
+            valid_tok = (attention_mask[..., :K] == 0)  # broadcastable to [B,1,Q,K]
+            page_any_valid = self._page_any_valid(valid_tok, page_size, num_pages)  # [B,Q,P]
+            page_any_valid = page_any_valid.unsqueeze(1).expand(B, H, Q, num_pages)  # [B,H,Q,P]
+            page_scores = torch.where(
+                page_any_valid, page_scores, torch.finfo(page_scores.dtype).min
+            )
+
+        # Step 3: choose Kp pages per (B,H,Q)
+        k_pages = min(num_pages, max(3, heavy_tokens // page_size))
+        if k_pages <= 0:
+            k_pages = 1
+        topk_pages = torch.topk(page_scores, k=k_pages, dim=-1, largest=True).indices  # [B,H,Q,Kp]
+
+        # Start from previous dense mask
+        dense_mask = previous_mask.get_dense_mask()
```

Comment:
previous mask is not a 0 - 1 mask . it can have values in [0,1] . where value represents the probability with which it is used. I think your handling here assumes binary masks.